### PR TITLE
graphfetch parallelize

### PR DESCRIPTION
**Pull request by tanujgirdhar**
#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/graphFetch/relationalGraphFetch.pure [hunk 1]
```diff
@@ -121,7 +121,7 @@ function <<access.private>> meta::relational::graphFetch::executionPlan::getTupl
    if($dbType == DatabaseType.DB2,
     | 100,
     | if($dbType == DatabaseType.Snowflake,
-      | 1000,
```
Comment (gs-ssh16) [inline legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/graphFetch/relationalGraphFetch.pure:124]: Should we even change this now? Ideally, we should determine it by the SQL stmt size / list size

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/GraphObjectsBatch.java [hunk 2]
```diff
@@ -68,35 +134,46 @@ public List<?> getObjectsForNodeIndex(int index)
...
+            return;
+        }
         this.xStorePropertyCaches.put(index, cache);
     }
 
     public ExecutionCache<GraphFetchCacheKey, List<Object>> getXStorePropertyCacheForNodeIndex(int index)
     {
-        return this.xStorePropertyCaches.get(index);
+        ExecutionCache<GraphFetchCacheKey, List<Object>> cache = this.xStorePropertyCaches.get(index);
+        if (cache != null && cache.equals(nullCache))
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/GraphObjectsBatch.java:149]: Can we simplify? -> nullCache.equals(cache)

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResultWithExecInfo.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.finos.legend.engine.plan.execution.concurrent.ParallelGraphFetchExecutionExecutorPool;
+import org.finos.legend.engine.plan.execution.result.Result;
+import org.finos.legend.engine.plan.execution.result.ResultVisitor;
+
+import java.util.concurrent.Future;
+
+public class DelayedGraphFetchResultWithExecInfo extends Result
+{
+    public final Future<DelayedGraphFetchResult> delayedGraphFetchResultFuture;
+    public final boolean executedInNewThread;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResultWithExecInfo.java:26]: Should these be public?

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class ParallelGraphFetchExecutionConfig
+{
+    public final long DEFAULT_DEFAULT_MAX_CONNECTIONS_PER_DATABASE = 10;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java:25]: Two times DEFAULT

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    public final long DEFAULT_PARALLEL_EXECUTION_POOL_SIZE = 100;
+
+    @JsonProperty
+    public final long parallelExecutionPoolSize;
+    @JsonProperty
+    private final long defaultMaxConnectionsPerDatabase;
+    @JsonProperty
+    private final Map<String, Integer> maxConnectionsPerDatabaseTypeOverrides;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java:33]: Database concepts cannot come into this space. You should introduce a relational extension concept

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/GraphFetchExecutionConfiguration.java [hunk 6]
```diff
@@ -33,13 +36,29 @@ public class GraphFetchExecutionConfiguration
...
+        this.parallelGraphFetchExecutionConfig = new ParallelGraphFetchExecutionConfig();
+    }
+
+    public GraphFetchExecutionConfiguration(ParallelGraphFetchExecutionConfig parallelGraphFetchExecutionConfig)
+    {
+        this.batchMemoryLimit = DEFAULT_BATCH_MEMORY_LIMIT;
+        this.softMemoryLimitPercentage = DEFAULT_SOFT_MEMORY_LIMIT_PERCENTAGE;
+        this.useAdaptiveBatching = DEFAULT_USE_ADAPTIVE_BATCHING;
+        this.defaultBatchSize = DEFAULT_BATCH_SIZE;
+        this.parallelizeGraphfetchQueries = true;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/GraphFetchExecutionConfiguration.java:60]: f should be capital in Graphfetch

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResult.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import org.finos.legend.engine.plan.execution.result.ResultVisitor;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+public class DelayedGraphFetchResult extends Result
+{
+    private final Runnable parentAdder;
+    private final Callable<List<DelayedGraphFetchResultWithExecInfo>> childVisitor;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResult.java:27]: Should it be called childExecutor?

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResultWithExecInfo.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        super("success");
+        this.threadConnectionKey = threadConnectionKey;
+        this.delayedGraphFetchResultFuture = delayedGraphFetchResultFuture;
+        this.executedInNewThread = executedInNewThread;
+    }
+
+    @Override
+    public <T> T accept(ResultVisitor<T> resultVisitor)
+    {
+        return null;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResultWithExecInfo.java:41]: throw exception

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResult.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        super("success");
+        this.parentAdder = parentAdder;
+        this.childVisitor = executeTempTableChild;
+    }
+
+    @Override
+    public <T> T accept(ResultVisitor<T> resultVisitor)
+    {
+        return null;
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/result/graphFetch/DelayedGraphFetchResult.java:39]: throw exception

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/concurrent/ParallelGraphFetchExecutionExecutorPool.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.Semaphore;
+
+public final class ParallelGraphFetchExecutionExecutorPool implements AutoCloseable
```
Comment (gs-ssh16) [inline legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/concurrent/ParallelGraphFetchExecutionExecutorPool.java:30]: Should the number of used threads be available as a metric to monitor? In general, pool state

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java (hunk 4 above)
Comment (tanujgirdhar) [reply legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java]: that's deliberate.

File: legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java (hunk 5 above)
Comment (tanujgirdhar) [reply legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/graphFetch/ParallelGraphFetchExecutionConfig.java]: moved relational configuration into https://github.com/finos/legend-engine/pull/2077/files#diff-e8404bcc8de985e707449918d183939695512ba0fdcab75c1e90089a2ab48d29. 
pool is also store agnostic: https://github.com/finos/legend-engine/pull/2077/files#diff-48aee8e3116e7f91188d4476f182eb43ba061092ff6d86cd53f14adfeb0a19b6.

**Comment by gs-ssh16** [comment]
I think we can go ahead with this change for now. May be revisit to see what can be improved before turning on parallel execution by default.

**Review by gs-ssh16 — APPROVED** [review]

**Review by epsstan — APPROVED** [review]

**Pull request merged**